File: lib/services/announcement_read_service.dart
import 'package:shared_preferences/shared_preferences.dart';

import '../models/rt_unit_model.dart';

/// Status baca pengumuman per RT (lokal di perangkat).
class AnnouncementReadService {
  AnnouncementReadService._();
  static final AnnouncementReadService instance = AnnouncementReadService._();

  static const _prefsKey = 'announcement_reads_v1';

  Future<Set<String>> _loadAll() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getStringList(_prefsKey)?.toSet() ?? {};
  }

  Future<void> _saveAll(Set<String> ids) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList(_prefsKey, ids.toList());
  }

  String _key(String rtId, String announcementId) => '$rtId:$announcementId';

  Future<bool> isRead(String rtId, String announcementId) async {
    final all = await _loadAll();
    return all.contains(_key(rtId, announcementId));
  }

  Future<void> markRead(String rtId, String announcementId) async {
    final all = await _loadAll();
    all.add(_key(rtId, announcementId));
    await _saveAll(all);
  }

  /// Pengumuman yang relevan untuk badge: pinned atau 14 hari terakhir, belum dibaca.
  Future<int> unreadCount(String rtId, List<AnnouncementModel> items) async {
    if (items.isEmpty) return 0;
    final all = await _loadAll();
    final since = DateTime.now().subtract(const Duration(days: 14));
    var count = 0;
    for (final a in items) {
      if (all.contains(_key(rtId, a.id))) continue;
      if (a.isPinned || a.createdAt.isAfter(since)) count++;
    }
    return count;
  }

  /// ID pengumuman yang sudah dibaca di RT ini.
  Future<Set<String>> readIdsForRt(String rtId) async {
    final all = await _loadAll();
    final prefix = '$rtId:';
    return all
        .where((k) => k.startsWith(prefix))
        .map((k) => k.substring(prefix.length))
        .toSet();
  }

  Future<List<AnnouncementModel>> filterUnread(String rtId, List<AnnouncementModel> items) async {
    final all = await _loadAll();
    final since = DateTime.now().subtract(const Duration(days: 14));
    return items.where((a) {
      if (all.contains(_key(rtId, a.id))) return false;
      return a.isPinned || a.createdAt.isAfter(since);
    }).toList();
  }
}

File: mobile/src/services/announcementReadService.ts
// Port dari lib/services/announcement_read_service.dart (SharedPreferences → AsyncStorage)
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Announcement } from '../types/models';

const PREFS_KEY = 'announcement_reads_v1';
const key = (rtId: string, id: string) => `${rtId}:${id}`;

async function loadAll(): Promise<Set<string>> {
  const raw = await AsyncStorage.getItem(PREFS_KEY);
  if (!raw) return new Set();
  try {
    return new Set(JSON.parse(raw) as string[]);
  } catch {
    return new Set();
  }
}

async function saveAll(ids: Set<string>): Promise<void> {
  await AsyncStorage.setItem(PREFS_KEY, JSON.stringify([...ids]));
}

const FOURTEEN_DAYS_MS = 14 * 24 * 60 * 60 * 1000;

export const announcementReadService = {
  async isRead(rtId: string, id: string): Promise<boolean> {
    return (await loadAll()).has(key(rtId, id));
  },

  async markRead(rtId: string, id: string): Promise<void> {
    const all = await loadAll();
    all.add(key(rtId, id));
    await saveAll(all);
  },

  /** Pengumuman relevan untuk badge: pinned atau ≤14 hari, belum dibaca. */
  async unreadCount(rtId: string, items: Announcement[]): Promise<number> {
    if (items.length === 0) return 0;
    const all = await loadAll();
    const since = Date.now() - FOURTEEN_DAYS_MS;
    let count = 0;
    for (const a of items) {
      if (all.has(key(rtId, a.id))) continue;
      if (a.isPinned || a.createdAt.getTime() > since) count++;
    }
    return count;
  },

  async readIdsForRt(rtId: string): Promise<Set<string>> {
    const all = await loadAll();
    const prefix = `${rtId}:`;
    return new Set([...all].filter((k) => k.startsWith(prefix)).map((k) => k.slice(prefix.length)));
  },

  async filterUnread(rtId: string, items: Announcement[]): Promise<Announcement[]> {
    const all = await loadAll();
    const since = Date.now() - FOURTEEN_DAYS_MS;
    return items.filter((a) => {
      if (all.has(key(rtId, a.id))) return false;
      return a.isPinned || a.createdAt.getTime() > since;
    });
  },
};
